### PR TITLE
Add DSP40

### PR DESCRIPTION
**Pull request by nooges**
## Description

Adds DSP40 keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/keebio/dsp40/config.h [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (drashna) [inline keyboards/keebio/dsp40/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/dsp40.c [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "dsp40.h"
```
Comment (drashna) [inline keyboards/keebio/dsp40/dsp40.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/dsp40.h [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#pragma once
```
Comment (drashna) [inline keyboards/keebio/dsp40/dsp40.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/keymaps/bakingpy/config.h [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/bakingpy/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c [hunk 5]
```diff
@@ -0,0 +1,200 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+extern keymap_config_t keymap_config;
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c:3]: Already included.
```suggestion
```

File: keyboards/keebio/dsp40/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,98 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+#define _QWERTY 0
+#define _LOWER 1
+#define _RAISE 2
+#define _ADJUST 3
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/default/keymap.c:6]:
```suggestion
enum custom_layer {
    _QWERTY,
    _LOWER,
    _RAISE,
    _ADJUST,
};
```

File: keyboards/keebio/dsp40/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,57 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/via/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+#define _MAIN 0
+#define _FN1 1
+#define _FN2 2
+#define _FN3 3
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/via/keymap.c:6]:
```suggestion
enum via_layers {
    _MAIN,
    _FN1,
    _FN2,
    _FN3,
};
```

File: keyboards/keebio/dsp40/rev1/rev1.c [hunk 11]
```diff
@@ -0,0 +1,10 @@
+#include "rev1.h"
```
Comment (drashna) [inline keyboards/keebio/dsp40/rev1/rev1.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/keebio/dsp40/rev1/rules.mk [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no 		# Enable support for HD44780 based LCDs
+ENCODER_ENABLE = yes
```
Comment (drashna) [inline keyboards/keebio/dsp40/rev1/rules.mk:24]:
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
ENCODER_ENABLE = yes
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+#define KC_LOWR LOWER
+#define KC_RASE RAISE
+#define KC_GRVF LT(_FKEYS, KC_GRV)
+#define KC_ENTS MT(MOD_LSFT, KC_ENT)
+#define KC_BL_S BL_STEP
+#define KC_BL_T BL_TOGG
+#define KC_RMOD RGB_MOD
+
+#ifndef LAYOUT_kc_ortho_4x12
+#define LAYOUT_kc_ortho_4x12( \
```
Comment (drashna) [inline keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c:54]: /cries

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Lint is mad.

File: keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c (hunk 13 above)
Comment (nooges) [reply keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c:54]: Yeah, one day I'll kill this all off

File: keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |             |      |      |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+  [_ADJUST] = LAYOUT_ortho_4x12( \
+    _______, RESET  , RGB_TOG, RGB_MOD, RGB_HUD, RGB_HUI,          RGB_SAD, RGB_SAI, RGB_VAD, RGB_VAI, _______, _______, \
+    _______, _______, _______, AU_ON,   AU_OFF,  AG_NORM,          AG_SWAP, MAC,     WINDOWS, TESTMODE,_______, _______, \
+    _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______, \
+    _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______ \
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/keymaps/bakingpy/keymap.c:158]:
```suggestion
  [_ADJUST] = LAYOUT_ortho_4x12(
    _______, RESET  , RGB_TOG, RGB_MOD, RGB_HUD, RGB_HUI,          RGB_SAD, RGB_SAI, RGB_VAD, RGB_VAI, _______, _______,
    _______, _______, _______, AU_ON,   AU_OFF,  AG_NORM,          AG_SWAP, MAC,     WINDOWS, TESTMODE,_______, _______,
    _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______,
    _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______
```

File: keyboards/keebio/dsp40/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+     * | Esc|  Q |  W |  E |  R |  T |  Y |  U |  I |  O |  P | BS |
+     * |-----------------------------------------------------------|
+     * | Tab |  A |  S |  D |  F |  G |  H |  J |  K |  L | Ent    |
+     * |-----------------------------------------------------------|
+     * | LSft   |  Z |  X |  C |  V |  B |  N |  M |  ,  |  . | /? |
+     * |-----------------------------------------------------------|
+     * | LCtl | LGui| LAlt|  spc fn0  |  spc fn1    |fn2|RAlt|RCtl |
+     * `-----------------------------------------------------------'
+     */
+    [_QWERTY] = LAYOUT( \
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/keymaps/default/keymap.c:45]:
```suggestion
    [_QWERTY] = LAYOUT(
```

File: keyboards/keebio/dsp40/keymaps/default/keymap.c [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        KC_ESC,  KC_Q,   KC_W,   KC_E,   KC_R,   KC_T,   KC_Y,   KC_U,   KC_I,    KC_O,    KC_P,   KC_BSPC,
+        KC_TAB,  KC_A,   KC_S,   KC_D,   KC_F,   KC_G,   KC_H,   KC_J,   KC_K,    KC_L,    KC_ENT,
+        KC_LSFT, KC_Z,   KC_X,   KC_C,   KC_V,   KC_B,   KC_N,   KC_M,   KC_COMM, KC_DOT,  KC_SLSH,
+        KC_LCTL, KC_LGUI, KC_LALT, LT(_LOWER, KC_SPC), LT(_RAISE, KC_SPC), LT(_ADJUST, KC_LGUI), KC_RALT, KC_RCTL
+    ),
+
+    [_LOWER] = LAYOUT( \
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/keymaps/default/keymap.c:52]:
```suggestion
    [_LOWER] = LAYOUT(
```

File: keyboards/keebio/dsp40/keymaps/default/keymap.c [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        KC_TILD, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, KC_BSPC,
+        KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_UNDS, KC_PLUS, KC_LCBR, KC_RCBR, KC_PIPE,
+        KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,S(KC_NUHS),S(KC_NUBS),BL_TOGG, BL_INC, BL_DEC,
+        _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_UP, KC_RGHT
+    ),
+
+    [_RAISE] = LAYOUT( \
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/keymaps/default/keymap.c:59]:
```suggestion
    [_RAISE] = LAYOUT(
```

File: keyboards/keebio/dsp40/keymaps/default/keymap.c [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+        KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_BSPC,
+        KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS,
+        KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_NUHS, KC_NUBS, _______, _______, KC_DEL,
+        _______, KC_TRNS, _______, KC_TRNS, KC_TRNS, _______, _______, RGB_TOG
+    ),
+
+    [_ADJUST] = LAYOUT( \
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/keymaps/default/keymap.c:66]:
```suggestion
    [_ADJUST] = LAYOUT(
```

File: keyboards/keebio/dsp40/readme.md [hunk 19]
```diff
@@ -0,0 +1,13 @@
+# DSP40
+
+A 40% keyboard with ortho and staggered layout options.
+
+Keyboard Maintainer: [Bakingpy/nooges](https://github.com/nooges)  
+Hardware Supported: STM32F072CBT6  
+Hardware Availability: [Keebio](https://keeb.io/)
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/readme.md:7]:
```suggestion
* Keyboard Maintainer: [Bakingpy/nooges](https://github.com/nooges)
* Hardware Supported: STM32F072CBT6
* Hardware Availability: [Keebio](https://keeb.io/)
```

File: keyboards/keebio/dsp40/rev1/config.h [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xCB10
+#define PRODUCT_ID      0x144C
+#define DEVICE_VER      0x0100
+#define MANUFACTURER    Keebio
+#define PRODUCT         DSP40 Rev. 1
+#define DESCRIPTION     40% Keyboard with identity crisis
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/rev1/config.h:27]:
```suggestion
```

File: keyboards/keebio/dsp40/rev1/readme.md [hunk 21]
```diff
@@ -0,0 +1,13 @@
+# DSP40 Rev. 1
+
+A 40% keyboard with ortho and staggered layout options.
+
+Keyboard Maintainer: [Bakingpy/nooges](https://github.com/nooges)  
+Hardware Supported: STM32F072CBT6  
+Hardware Availability: [Keebio](https://keeb.io/)
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/rev1/readme.md:7]:
```suggestion
* Keyboard Maintainer: [Bakingpy/nooges](https://github.com/nooges)
* Hardware Supported: STM32F072CBT6
* Hardware Availability: [Keebio](https://keeb.io/)
```

File: keyboards/keebio/dsp40/rev1/rules.mk [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
```
Comment (fauxpark) [inline keyboards/keebio/dsp40/rev1/rules.mk:19]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**